src/types.rs

```diff
@@ -218,3 +220,74 @@ impl LintPass for CastPass {
         }
     }
 }
+
+declare_lint!(pub TYPE_COMPLEXITY, Warn,
+              "usage of very complex types; recommends factoring out parts into `type` definitions");
+
+#[allow(missing_copy_implementations)]
+pub struct TypeComplexityPass(pub NodeSet);
+
+impl TypeComplexityPass {
+    pub fn new() -> TypeComplexityPass {
+        TypeComplexityPass(NodeSet())
+    }
+}
+
+impl LintPass for TypeComplexityPass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(TYPE_COMPLEXITY)
+    }
+
+    fn check_ty(&mut self, cx: &Context, ty: &ast::Ty) {
+        if self.0.contains(&ty.id) {
```

Comment:
Instead of maintaining a alloc-heavy hashmap, why not use the existing node map and ensure that the parent node (via http://manishearth.github.io/rust-internals-docs/rustc/ast_map/struct.Map.html#method.get_parent_node) isn't a `Ty`?